drivers:iio:dac: add max538x support

mrcsosa opened the pull request:
Added driver implementation for MAX5380/1/2 8-bit DAC.

More info: analog.com: MAX5380-MAX5382

Signed-off-by: Marc Paolo Sosa [marcpaolo.sosa@analog.com](mailto:marcpaolo.sosa@analog.com)

amiclaus reviewed (commented):
please send the patches upstream for further review.

mrcsosa commented:
Hi, can I request an initial review of this PR before sending the patches upstream? Thank you.

nunojsa commented on Documentation/devicetree/bindings/iio/dac/maxim,max5380.yaml:
```diff
@@ -0,0 +1,51 @@
+    const: 1
+
+  '#size-cells':
+    const: 0
```
the above two properties are not needed... On top of this, if you are not adding more into this schema, the part could go to:

https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/trivial-devices.yaml

That said, you could already add support for the vdd regulator that would justify this bindings.

nunojsa commented on Documentation/devicetree/bindings/iio/dac/maxim,max5380.yaml:
```diff
@@ -0,0 +1,51 @@
+         #address-cells = <1>;
+         #size-cells = <0>;
+
+         max5380: dac@31 {
```
just `dac@31`

nunojsa commented on drivers/iio/dac/Kconfig:
```diff
@@ -371,6 +371,18 @@ config MAX517
+config MAX5380
+	tristate "Maxim MAX5380 DAC driver"
+	depends on I2C
+	depends on OF
```
there's no OF dependency...

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+#include <linux/iio/iio.h>
+#include <linux/iio/sysfs.h>
+
+#define V_REF 2.0
```
use something like `define MAX5380_VREF 2`

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+#include <linux/proc_fs.h>
+#include <linux/i2c.h>
+#include <linux/iio/iio.h>
+#include <linux/iio/sysfs.h>
```
please cleanup the includes... There are things not needed and others that are missing. Furthermore, they should be placed in alphabetical order

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+struct max5380_data {
+	struct i2c_client *client;
+	u8 value;
+};
```
I do not think we need any specific struct for such a simple part... see below

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+	if (mask == IIO_CHAN_INFO_RAW) {
+		if (val < 0 || val > 255)
+			return -EINVAL;
+		data->value = val;
```
no need to cache this value (I think). Also a note that you should only cache the value in case i2c_smbus_write_byte() succeeds

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+		}
+	} else {
+		return -EINVAL;
+	}
```
please still do a  switch() statement. It's more coherent with what is done in IIO

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+
+static int max5380_read_raw(struct iio_dev *indio_dev,
+							struct iio_chan_spec const *chan,
+							int *val, int *val2, long mask)
```
take care about coding style

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+	struct max5380_data *data = iio_priv(indio_dev);
+
+	if (mask == IIO_CHAN_INFO_RAW) {
+		*val = data->value;
```
Can't we read the value from the device? Is the only way this is acceptable (IMO)

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+		return -EINVAL;
+	}
+
+	return IIO_VAL_FRACTIONAL;
```
return in the correct switch() block

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+	data = iio_priv(indio_dev);
+	data->client = client;
+
+	indio_dev->dev.parent = &client->dev;
```
drop this

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+	return ret;
+	}
+
+return 0;
```
wrong indentation... furthermore, just do `return devm_iio_device_register()`. Note that you need the devm_ variant. Otherwise you would need a .remove hook defined in your `struct i2c_driver`

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+
+static const struct i2c_device_id max5380_id[] = {
+	{ "max5380", 0 },
+	{},
```
no need to explicitly set .driver_data to 0.

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+static struct i2c_driver max5380_driver = {
+		.driver = {
+				.name = "max5380",
+				.of_match_table = max5380_of_match,
```
this looks to have more indentation than needed...

nunojsa commented on drivers/iio/dac/max5380.c:
```diff
@@ -0,0 +1,128 @@
+		if (val < 0 || val > 255)
+			return -EINVAL;
+		data->value = val;
+		ret = i2c_smbus_write_byte(data->client, data->value);
```
why smbus? from a quick grasp on the datasheet, I do not see any specific smbus mention... Can't we use [i2c_master_send()](https://elixir.bootlin.com/linux/latest/C/ident/i2c_master_send)?

mrcsosa replied:
Hi, I have added few updates to support other devices. I will also include the improvements you have mentioned on my next update. Thank you so much.

nunojsa reviewed (changes requested):
I would suggest to make this driver a bit more interesting. The part is very simple and looking at the datasheet, I would say the driver should support all max5380 - max5382. That would turn the driver a bit more interesting. If I'm not missing something, you would only need to handle the different VREFs

nunojsa commented on Documentation/devicetree/bindings/iio/dac/maxim,max538x.yaml:
```diff
@@ -0,0 +1,62 @@
+    const: 1
+
+  '#size-cells':
+    const: 0
```
no need for the above two properties..

nunojsa commented on Documentation/devicetree/bindings/iio/dac/maxim,max538x.yaml:
```diff
@@ -0,0 +1,62 @@
+      - maxim,max5382L
+      - maxim,max5382M
+      - maxim,max5382N
+      - maxim,max5382K
```
What is the L,M,N,K,L all about? Do we need a sepcific compatible? AFAICT, we have  max5380/max5381/max53802

nunojsa commented on Documentation/devicetree/bindings/iio/dac/maxim,max538x.yaml:
```diff
@@ -0,0 +1,62 @@
+                reg=<0x31>;
+          };
+    };
+...
```
as it stands, these could go to trivial devices... But I think you need to add the vdd regulator if supporting all three devices

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+		return ret;
+	}
+
+	return 0;
```
just `return devm_iio_device_register()`

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+	case MAX5382M:
+	case MAX5382N:
+	case MAX5382K:
+		data->vfactor = 0.9;
```
? There's no floating point in the kernel... Also, I do not think this is right. From the datasheet, the internal reference for mx5832 is 0.9 * VDD. Hence, you need to use a regulator and get it's voltage and then do the calculation accordingly.

Look for examples using `devm_regulator_get()` and `regulator_get_voltage()`

You can have a flag (most obvious solution) in your chip info structure to let you know when do you need to get the regulator voltage so that you can calculate the internal reference.

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	indio_dev->dev.parent = &client->dev;
+	indio_dev->name = dev_name(&client->dev);
```
do not use `dev_name()`. If supporting multiple devices, you could have a chip_info strutuctue with a char *name and the vfactor. Then you could use those... The names should just be "max5380", "max5381" or "max5382". There's plenty of drivers using the "chip_info" approach. You can see an example in here:

https://elixir.bootlin.com/linux/latest/source/drivers/iio/imu/adis16475.c#L657

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+		return -EINVAL;
+	}
+
+	indio_dev->dev.parent = &client->dev;
```
drop this. automatically done by IIO core

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+
+	indio_dev = devm_iio_device_alloc(&client->dev, sizeof(*data));
+	if (!indio_dev) {
+		dev_err(&client->dev, "Failed to allocate IIO device\n");
```
remove this log

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+		dev_err(&client->dev, "Invalid device type specified\n");
+
+		return -EINVAL;
+	}
```
remove the above... should never happen

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+		}
+	} else {
+		return -EINVAL;
+	}
```
use a switch() case

nunojsa commented on drivers/iio/dac/max538x.c:
```diff
@@ -0,0 +1,218 @@
+		.name = "max538x",
+		.of_match_table = max538x_of_match,
+	},
+	.probe = max538x_probe,
```
use probe_new()... You need to see when sending these upstream if probe_new() is still in use or not. If it is, you should use it (I think).

mrcsosa commented:
Removed: Multiple device list (adjusted to only max5380/1/2)
Removed: Unnecessary error checking and logs
Adjusted: if statements to switch case
Added: voltage regulator for external vref.

nunojsa commented:
@mrcsosa, I completely failed to look again at this one... Can you rebase onto master?

For similar situations, just ping me for the review :)

nunojsa commented:
More than year olde. @mrcsosa what is the status on this one? Should we close?

mrcsosa commented:
Closing this pull request
